angr/engines/pcode/engine.py

```diff
@@ -0,0 +1,200 @@
+import claripy
+import logging
+
+from angr.engines.engine import SuccessorsMixin
+from ...utils.constants import DEFAULT_STATEMENT
+from ... import sim_options as o
+from ... import errors
+from .lifter import PcodeLifterEngineMixin, lifters
+from .emulate import PcodeEmulatorMixin
+
+l = logging.getLogger(__name__)
+
+# pylint:disable=abstract-method
+
+class HeavyPcodeMixin(
+    SuccessorsMixin,
+    PcodeLifterEngineMixin,
+    PcodeEmulatorMixin,
+):
+    """
+    Execution engine based on P-code, Ghidra's IR.
+
+    Responds to the following parameters to the step stack:
+
+    - irsb:        The P-Code IRSB object to use for execution. If not provided one will be lifted.
+    - skip_stmts:  The number of statements to skip in processing
+    - last_stmt:   Do not execute any statements after this statement
+    - thumb:       Whether the block should be force to be lifted in ARM's THUMB mode. (FIXME)
+    - extra_stop_points:
+                   An extra set of points at which to break basic blocks
+    - insn_bytes:  A string of bytes to use for the block instead of the project.
+    - size:        The maximum size of the block, in bytes.
+    - num_inst:    The maximum number of instructions.
+    """
+
+    def process_successors(
+        self,
+        successors,
+        irsb=None,
+        insn_text=None,
+        insn_bytes=None,
+        thumb=False,
+        size=None,
+        num_inst=None,
+        extra_stop_points=None,
+        **kwargs
+    ): # pylint:disable=arguments-differ
+        if not lifters[self.state.arch.name] or type(successors.addr) is not int:
+            return super().process_successors(successors, extra_stop_points=extra_stop_points, num_inst=num_inst, size=size, insn_text=insn_text, insn_bytes=insn_bytes, **kwargs)
+
+        if insn_text is not None:
+            if insn_bytes is not None:
+                raise errors.SimEngineError(
+                    "You cannot provide both 'insn_bytes' and 'insn_text'!"
+                )
+
+            insn_bytes = self.project.arch.asm(
+                insn_text, addr=successors.addr, thumb=thumb
+            )
+            if insn_bytes is None:
+                raise errors.AngrAssemblyError(
+                    "Assembling failed. Please make sure keystone is installed, and the"
+                    " assembly string is correct."
+                )
+
+        successors.sort = "IRSB"
+        successors.description = "IRSB"
+        self.state.history.recent_block_count = 1
+        self.state.scratch.guard = claripy.true
+        self.state.scratch.sim_procedure = None
+        addr = successors.addr
+        self.state.scratch.bbl_addr = addr
+
+        while True:
+            if irsb is None:
+                irsb = self.lift_vex(
```

Comment:
Rename to `lift_pcode`

Reply:
Renamed to `lift_pcode`, added `lift_vex` compatibility interface for integration.